### ISSUE TITLE
Multiple errors reporting

### ISSUE THREAD

**Issue by rabidpraxis** [honeybadger-io/honeybadger-elixir#242]
We have had a few reports of 2 errors reporting. Based on the notice data, it seems like both the `Plug` error handler and the logger are reporting the same exception. It appears something like this has been a problem before (#52, #56) and was resolved.

I  have noticed this when an error is thrown in the an Ecto.Repo process (say an invalid query) then I see an error reported for the Repo process and also the Phoenix/Plug process.

**Comment by henrik** [comment]
Seeing the same. To give context on why one might turn on both the plug and the logger: I did so thinking the former would give better error reporting in the web stack and the latter would catch any non-web-stack errors. But of course it causes double reports a lot of the time.

**Comment by rabidpraxis** [comment]
Based on my digging, because the Logger and Plug process are unaware of each other (it's erlang), we don't have a great way to know if an error has been notified, so as to halt sending one of the notices.

It would be preferable to have the Plug take precedence over the Logger if they are both enabled, as the Plug notifier injects helpful metadata for debugging. I have seen other projects that look into the stack trace to and ignore if any pattern match from Cowboy (as that is a clue that plug would be used). I'm not a huge fan of that solution.

It's been a while since I have been in Elixir land but I had an idea around this. Could we:

- Update our `Honeybadger.Plug` to include an overridden `call` function
- Inside the call, utilize the new (as of 1.9) `Logger.metadata[:domain]` property by appending something like `:honeybadger_plug` to the domain list. 
- Ignore any error events in our `Honeybadger.Logger` if we see the `:honeybadger_plug` key in the domain?

**Comment by messutied** [comment]
After looking at `Honeybadger.Plug` and `Plug.ErrorHandler` not sure how what you suggest could work, since in `Plug.ErrorHandler` what's happening is a re-raising of the exception, no `Logger` being used to log it (as far as I understand).

I found this issue interesting https://github.com/elixir-plug/plug/issues/444, since it's from 2016 perhaps Rollbax already found a workaround?

And finally I was thinking that perhaps this could be more easily solved on your backend, you could look at errors logged and if parameters `kind, reason, stack` are identical within some threshold like for a plug logged error and a Logger logged error automatically remove the Logger logged error.

**Comment by rabidpraxis** [comment]
> After looking at Honeybadger.Plug and Plug.ErrorHandler not sure how what you suggest could work, since in Plug.ErrorHandler what's happening is a re-raising of the exception, no Logger being used to log it (as far as I understand).

Hello again Eduardo, I appreciate you spending time looking into this!

We call `notify` in the plug:

https://github.com/honeybadger-io/honeybadger-elixir/blob/634c68e783576acb80eeb327a0caee2618340f08/lib/honeybadger/plug.ex#L63-L67

And in the Logger backend:

https://github.com/honeybadger-io/honeybadger-elixir/blob/634c68e783576acb80eeb327a0caee2618340f08/lib/honeybadger/logger.ex#L29-L47

The Logger module receives an `:error` message from the bubbled up error in the process because we added it as a backend:

https://github.com/honeybadger-io/honeybadger-elixir/blob/634c68e783576acb80eeb327a0caee2618340f08/lib/honeybadger.ex#L197-L199

So the idea is if we know that the plug is being used, then we would exit before notifying in the `Honeybadger.Logger` handle_event call.

**Comment by messutied** [comment]
Hi @rabidpraxis, I understand the idea, I just don't see how to communicate from the Plug to the Logger backend, you mentioned overriding the `call` function before, I suppose you mean overriding `Plug.ErrorHandler.call/2` in `Honeybadger.Plug`? I don't think that's possible :(

**Comment by rabidpraxis** [comment]
@messutied The hope was to add another plug, something similar to the internals of `Plug.ErrorHandler`: 

https://github.com/elixir-plug/plug/blob/692655393a090fbae544f5cd10255d4d600e7bb0/lib/plug/error_handler.ex#L59-L83

I attempted to do that, and while the plug worked, my updates to the logger metadata did not propagate to the duplicated errors. My assumption is that those errors are coming from cowboy itself and the metadata that I am adding is not in the same process.

I ended up getting the duplicates to stop by ignoring all error events from cowboy in the Logger (which have [:cowboy]) as its logger domain. Since cowboy is most often used as the backend for Phoenix, it seems safe to remove those events as they should also be caught by the ErrorHandler.

 I added a PR ☝🏼 for that experiment.
